cumulusci/tasks/salesforce/content_documents.py

```diff
@@ -0,0 +1,209 @@
+            self.logger.info(f'    Visibility: "{visibility}"')
+            for record_id in record_ids:
+                self.sf.ContentDocumentLink.create(
+                    {
+                        "ContentDocumentId": content_document_id,
+                        "LinkedEntityId": record_id,
+                        "ShareType": share_type,
+                        "Visibility": visibility,
+                    }
+                )
+            self.logger.info(
+                f'Successfully linked {len(record_ids)} record{"" if len(record_ids) == 1 else "s"} to Content Document "{content_document_id}"'
+            )
+        else:
+            self.logger.info(
+                "😴 No records IDs queried. Skipping linking the Content Document to related records."
+            )
+
+    def _run_task(self):
+        # Insert the ContentDocument.
+        content_document_id = self._insert_content_document()
+
+        # Query records to link to the new ContentDocument.
+        # "Rolls back" the ContentDocument insert if an Exception is raised.
+        record_ids = None
+        get_record_ids_to_link_exception = None
+        try:
+            record_ids = self._get_record_ids_to_link()
+        except SalesforceMalformedRequest as e:
+            get_record_ids_to_link_exception = to_cumulusci_exception(e)
+        except Exception as e:
+            get_record_ids_to_link_exception = e
+        finally:
+            if get_record_ids_to_link_exception:
+                self.logger.error(
+                    "An error occurred querying records to link to the ContentDocument."
+                )
+                self.logger.error(
+                    f'Deleting ContentDocument "{content_document_id}" to roll back the transaction.'
+                )
+                self._delete_content_document(content_document_id)
+                # Reraise the Exception
+                raise get_record_ids_to_link_exception
+
+        # Links queried records to link to the new ContentDocument.
+        # "Rolls back" the ContentDocument insert if an Exception is raised.
+        link_records_to_content_document_exception = None
+        try:
+            self._link_records_to_content_document(content_document_id, record_ids)
+        except SalesforceMalformedRequest as e:
+            link_records_to_content_document_exception = to_cumulusci_exception(e)
+        except Exception as e:
+            link_records_to_content_document_exception = e
+        finally:
+            if link_records_to_content_document_exception:
+                self.logger.error(
+                    "An error occurred linking queried records to the ContentDocument."
+                )
+                self.logger.error(
+                    f'Deleting ContentDocument "{content_document_id}" to roll back the transaction.'
+                )
+                self._delete_content_document(content_document_id)
```

Comment:
I thought I was being helpful rolling back the entire transaction.  For scratch org development, it's not necessary.   Maybe it's useful in sandbox/production environments?   However, there's not really a "rollback" feature in any other cci task, so it could be best just to remove this functionality.

Reply:
I'd say take it out if there isn't a clear need for it. It adds complexity, and probably makes it harder to debug an error when it happens.

Reply:
My intent is `InsertContentDocument` can be extended for other uses cases.   For example, `UploadProfilePhoto` has reuses the same `_insert_content_document` method but has a different way to "link" the file to something (in this case  using the Connect API @ `"connect/user-profiles/{user_id}/photo"`.   I deduced that inserting a file always involves:

1. Inserting the Content Document
2. Determining which record(s) to "link" the file to (if any)
3. "Link" the records to the file.

Reply:
Sure, I think catching those exceptions and reporting them in a more helpful way makes sense -- just not removing the ContentDocument when that happens.

Reply:
Done.   I still have to review the tests, but the production code is changed.